genetic_algorithms/cli/solve.py

```diff
@@ -0,0 +1,130 @@
+@click.option('-p', '--problem_model', type=config_metadata['problem_model']['type'], help='Problem to be solved')
+@click.option('-b', '--benchmark_file', help='Benchmark file on which problem will be run')
+@click.option('-a', '--algorithm', type=config_metadata['algorithm']['type'], help='Algorithm with which problem will be solved')
+@click.option('-v', '--visualization', is_flag=True)
+def solve(config_file, **cli_options):
+    """
+    Solves a problem based on config file or cli options. 
+    Config file should contain same keys as defined in click.option decorator above.
+    """
+    options = read_options(config_file, cli_options)
+    solver = create_solver(options)
+    problem_model = create_problem_model(options)
+    algorithm = create_algorithm(problem_model, options)
+    solution = solver.solve(problem_model, algorithm)
+    click.echo(solution)
+
+
+def read_options(config_file_path: str, cli_options):
+    options = cli_options
+    if config_file_path:
+        with open(config_file_path, 'r') as config:
+            config = json.load(config)
+            options = {k: cli_options[k] or config.setdefault(
+                k, None) for k in cli_options}
+    for key in required_options.keys():
+        prompt_if_not_exists(options, key, required_options)
+    console.log("Initialized with options: ", options)
+    return options
+
+
+def prompt_if_not_exists(options, key: str, metadata=None):
+    if options.setdefault(key, None) is None:
+        prompt = f'Select {key.replace("_", " ")}'
+        option_metadata = metadata.setdefault(key, {})
+        options[key] = click.prompt(prompt,
+                                    type=option_metadata.setdefault(
+                                        'type', None),
+                                    default=option_metadata.setdefault(
+                                        'default', None)
+                                    )
+
+
+def create_solver(options):
+    options = options.setdefault('solver_config', {})
+    console.print("Configuring solver", style="bold blue")
+    config = populate_options_for_dataclass(options, SolverConfig)
+    return LocalSearchSolver(config)
+
+
+def populate_options_for_dataclass(options, dataclass: Type):
+    for field in fields(dataclass):
+        field_metadata = {field.name: {
+            'default': field.default
+        }}
+        prompt_if_not_exists(options, field.name, field_metadata)
+    return dataclass(**options)
+
+
+def create_problem_model(options):
+    return Model.problems[options['problem_model']].from_benchmark(
+        options['benchmark_file'])
+
+
+def create_algorithm(problem_model: Model, options) -> Union[Algorithm, AlgorithmWrapper]:
+    algorithm_type = Algorithm._algorithms[options['algorithm']]
+    config_type = signature(algorithm_type).parameters['config'].annotation
+    config = options.setdefault(algorithm_type.__name__, {})
+    config = populate_options_for_dataclass(config, config_type)
```

Comment:
Z tego jestem najbardziej dumny, bo to pozwala na wyciągnięcie z typehint'a na `__init__` typ dataklasy z konfiguracją i późniejsze jej wypełnienie parametrami